infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -708,16 +742,55 @@ private boolean clusterRoleChanged(TopologyDescriptor discoveredTopology) {
     private void onStandbyClusterAddRemove(TopologyDescriptor discoveredTopology) {
         log.debug("Standby Cluster has been added or removed");
 
-        // We only need to process new standby's if your role is of an ACTIVE cluster
-        if (localClusterDescriptor.getRole() == ClusterRole.ACTIVE && replicationManager != null && isLeader.get()) {
-            replicationManager.processStandbyChange(discoveredTopology);
+        if (discoveredTopology.getTopologyConfigId() !=
+            topologyDescriptor.getTopologyConfigId()) {
+            log.warn("Detected changes in the topology." +
+                "The new topology descriptor {} doesn't have the same " +
+                "topologyConfigId as the current one {}", discoveredTopology,
+                topologyDescriptor);
+        }
+        // We only need to process new sinks if the local cluster role is SOURCE
+        if (localClusterDescriptor.getRole() == ClusterRole.ACTIVE &&
+            replicationManager != null && isLeader.get()) {
+            Set<String> receivedStandbys =
+                discoveredTopology.getStandbyClusters().keySet();
+            Set<String> currentStandbys =
+                topologyDescriptor.getStandbyClusters().keySet();
+
+            Set<String> intersection = Sets.intersection(currentStandbys,
+                receivedStandbys);
+
+            Set<String> standbysToRemove = Sets.difference(currentStandbys,
+                receivedStandbys);
+
+            Set<String> standbysToAdd = Sets.difference(receivedStandbys,
+                currentStandbys);
+
+            for (String remoteClusterId : standbysToRemove) {
+                remoteClientToMetadataManagerMap.remove(remoteClusterId);
+            }
+            createMetadataManagers(standbysToAdd);
+            replicationContext.setTopology(discoveredTopology);
+            replicationManager.setContext(replicationContext);
```

Comment:
It seems to make more sense to update the context within the method processStandbyChange, so it is self contained, cause when you look at that method where it's receiving a new topology, one believes we are missing updating the context but it's done off track.

Reply:
makes sense, I can move it there.  updateRuntimeConfigId() is now contained there anyway.

I also think that replicationContext need not be an instance variable in DiscoveryService as it is only used to pass to the ReplicationManager.  But not planning to make the change in this PR.